### PR TITLE
Add pundit and basic authorization

### PR DESCRIPTION
**Pull request by pawelkozlik1**

### CODE REVIEW

File: app/policies/reservation_policy.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      if user.manager? || user.employee?
+        scope.all
+      else
+        scope.where(user: user)
+      end
+    end
+  end
+
+  def show?
+    owner_or_staff?
```
Comment (wmaciejak) [inline app/policies/reservation_policy.rb:15]: That kind of naming is much better than `authorized?`

File: app/controllers/cinema_halls_controller.rb [hunk 2]
```diff
@@ -4,17 +4,20 @@ class CinemaHallsController < ApplicationController
...
 
   def show
     cinema_hall = CinemaHall.find(params[:id])
+    authorize cinema_hall
     render json: cinema_hall
   end
 
   def create
     cinema_hall = CinemaHall.new(cinema_hall_params)
+    authorize cinema_hall
```
Comment (wmaciejak) [inline app/controllers/cinema_halls_controller.rb]: Like I noticed on call, you can also just put here:
```ruby
def create
  authorize CinemaHall
```
cause this means also
```ruby
def create
  authorize CinemaHall, :create? # name of action to authorize is automatically assumed from name of controller method
```

I guess that it's much cleaner way of solving this.

File: app/controllers/movies_controller.rb [hunk 3]
```diff
@@ -12,8 +12,8 @@ def show
   end
 
   def create
+    authorize Movie
```
Comment (wmaciejak) [inline app/controllers/movies_controller.rb:15]: Here you did it really good!

File: app/controllers/movies_controller.rb [hunk 4]
```diff
@@ -22,11 +22,13 @@ def create
   end
 
   def update
+    authorize Movie
     movie = Movies::UseCases::Update.new(params: movie_params, id: params[:id]).call
     render json: Movies::Representers::Single.new(movie).call
   end
 
   def destroy
+    authorize Movie
```
Comment (wmaciejak) [inline app/controllers/movies_controller.rb:31]: You've not authorizing particular object here for destroy and update methods?

File: app/controllers/reservations_controller.rb [hunk 5]
```diff
@@ -4,16 +4,18 @@ class ReservationsController < ApplicationController
...
   end
 
   def show
     reservation = Reservation.find(params[:id])
+    authorize reservation
     render json: reservation
   end
 
   def create
+    authorize Reservaiton
```
Comment (wmaciejak) [inline app/controllers/reservations_controller.rb]: Typo

File: app/controllers/reservations_controller.rb [hunk 6]
```diff
@@ -4,16 +4,18 @@ class ReservationsController < ApplicationController
   before_action :authenticate_user!
 
   def index
-    reservations = Reservation.all
+    reservations = policy_scope(Reservation).where(screening: params[:screening_id])
```
Comment (wmaciejak) [inline app/controllers/reservations_controller.rb]: So here, you are not showing list of all reservations, but only reservations for particular screening?

File: app/controllers/screenings_controller.rb [hunk 7]
```diff
@@ -4,17 +4,20 @@ class ScreeningsController < ApplicationController
...
 
   def show
     screening = Screening.find(params[:id])
+    authorize screening
     render json: screening
   end
 
   def create
     screening = Screening.new(screening_params)
+    authorize screening
```
Comment (wmaciejak) [inline app/controllers/screenings_controller.rb]: The same comment like in CinemaHalls Controller

File: app/policies/cinema_hall_policy.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+class CinemaHallPolicy < ApplicationPolicy
+  attr_reader :user, :cinema_hall
+
+  def initialize(user, cinema_hall)
+    super(user)
+    @cinema_hall = cinema_hall
+  end
```
Comment (wmaciejak) [inline app/policies/cinema_hall_policy.rb]: Consider [this](https://github.com/atytko/monte-cinema/pull/29#discussion_r699987474) comment

File: app/controllers/movies_controller.rb (hunk 4 above)
Comment (pawelkozlik1) [reply app/controllers/movies_controller.rb:31]: I was thinking that I could just check if user is a manager or employee no matter which movie they want to update / delete, but I'm still not sure how to authorize an object called by UseCase with correct syntax. Would it be something like `authorize Movies::UsceCases::Update.new`? or the whole object like in line 26?

Comment (wmaciejak) [reply app/controllers/movies_controller.rb:31]: >I was thinking that I could just check if user is a manager or employee no matter which movie they want to update / delete

This is fine anyway 👌 

>but I'm still not sure how to authorize an object called by UseCase with correct syntax. 

IMO you can fetch here movie to validate if movie exist and then authorize if use have access to this movie, and then remove movie if everything is fine

**Pull request merged**